src/provider/mod.rs

```diff
@@ -405,21 +405,23 @@ impl RpcHandler {
         futures::stream::iter(items)
     }
     async fn provide(self, msg: ProvideRequest) -> anyhow::Result<ProvideResponse> {
-        let path = msg.path;
-        let data_sources: Vec<DataSource> = if path.is_dir() {
-            let mut paths = Vec::new();
-            let mut iter = tokio::fs::read_dir(&path).await?;
-            while let Some(el) = iter.next_entry().await? {
-                if el.path().is_file() {
-                    paths.push(el.path().into());
-                }
-            }
-            paths
-        } else if path.is_file() {
-            vec![path.into()]
-        } else {
-            anyhow::bail!("path must be either a Directory or a File");
-        };
+        let root = msg.path;
+        anyhow::ensure!(
+            root.is_dir() || root.is_file(),
+            "path must be either a Directory or a File"
+        );
+        let files = read_dir_recursive(&root)?;
```

Comment:
lets make this async? or at least call `spawn_blocking`

Reply:
I guess I should.

But I am getting a bit confused with all these branches building on persistence. I think we should merge that first.

Reply:
Stupid of me to not just use WalkDir. Have changed it.